src/network/netdev/l2tp-tunnel.c

```diff
@@ -0,0 +1,500 @@
+}
+
+static int netdev_l2tp_tunnel_create(NetDev *netdev) {
+        _cleanup_(sd_netlink_message_unrefp) sd_netlink_message *m = NULL;
+        L2tpSession *session = NULL;
+        uint32_t serial;
+        L2tpTunnel *t;
+        int r;
+
+        assert(netdev);
+
+        t = L2TP(netdev);
+
+        assert(t);
+
+        r = netdev_l2tp_fill_message_tunnel(netdev, &m);
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_send(netdev->manager->genl, m, &serial);
+        if (r < 0 && r != -EEXIST)
+                return log_netdev_error_errno(netdev, r, "Failed to create L2TP tunnel: %m");
+
+        LIST_FOREACH(sessions, session, t->sessions) {
+                _cleanup_(sd_netlink_message_unrefp) sd_netlink_message *n = NULL;
+
+                r = netdev_l2tp_fill_message_session(netdev, session, &n);
+                if (r < 0)
+                        return r;
+
+                r = sd_netlink_send(netdev->manager->genl, n, &serial);
+                if (r < 0 && r != -EEXIST)
+                        return log_netdev_error_errno(netdev, r, "Failed to add L2TP session: %m");
+
+        }
+
+        return 0;
+}
+
+int config_parse_l2tp_tunnel_address(const char *unit,
+                                     const char *filename,
+                                     unsigned line,
+                                     const char *section,
+                                     unsigned section_line,
+                                     const char *lvalue,
+                                     int ltype,
+                                     const char *rvalue,
+                                     void *data,
+                                     void *userdata) {
+        L2tpTunnel *t = userdata;
+        union in_addr_union *addr = data, buffer;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        if (t->family == AF_UNSPEC)
+                r = in_addr_from_string_auto(rvalue, &t->family, addr);
+        else
+                r = in_addr_from_string(t->family, rvalue, addr);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r,
+                           "Invalid L2TP Tunnel address specified in %s='%s', ignoring assignment: %m", lvalue, rvalue);
+                return 0;
+        }
+
+        *addr = buffer;
+
+        return 0;
+}
+
+int config_parse_l2tp_tunnel_port(const char *unit,
+                                  const char *filename,
+                                  unsigned line,
+                                  const char *section,
+                                  unsigned section_line,
+                                  const char *lvalue,
+                                  int ltype,
+                                  const char *rvalue,
+                                  void *data,
+                                  void *userdata) {
+        L2tpTunnel *t = userdata;
```

Comment:
this isn't right. the gperf table shows that this ptr is actually already of the port value itself, i.e.

```c
uint16_t *port = userdata;
```

would be correct....